Add KTX2 support and remove KTX/CRN

ebogo1 opened the pull request:
This PR is very similar to https://github.com/CesiumGS/cesium/pull/9040/. There are a few big to-do items before this is ready, but eslint passes and the KTX2 balloon in the "3D Models" sandcastle works so I figured it was a good time to get the code out there again. 

Changes include: 
- Remove `loadCRN` and `loadKTX` and replace with `loadKTX2`
    - rework `loadKTXSpec` -> `loadKTX2Spec`
- Add [Binomial transcoder](https://github.com/BinomialLLC/basis_universal) (basis_transcoder.wasm) 
    - Binomial module is used to parse KTX2 files in `parseKTX2.js`, which is called from the `transcodeKTX2` worker's `transcode()` function
- Updated Image-Based Lighting, 3D Models, and Materials Sandcastles with .ktx2 data

To-do: 
- [x] ~Investigate alternative transcoder module - https://github.com/KhronosGroup/Universal-Texture-Transcoders~ (this will be a follow-up PR)
- [x] UASTC/ASTC/BC7 support
- [x] Update sample data glTFs
- [x] Support for compressed .ktx2 with embedded mipmap (needs ModelSpec test)
    - glTF model with compressed + mipmap KTX2: [ktx2-box.zip](https://github.com/CesiumGS/cesium/files/6656581/ktx2-box.zip)
- [x] Update IBL instructions in Sandcastle
- [x] Add check for SFLOAT vs UNORM uncompressed formats
- [x] Finish specs for `loadKTX2`/`KTX2Transcoder`
- [x] Fix Image-Based Lighting 
- [x] Fix Materials sandcastles (ktx2 pixel format not supported)
- [x] Remove unused transcoder modules
- [x] Update specs using .ktx2 sample data
- [x] Test Sandcastles on different OS 
    - [x] iOS
    - [x] Android (Chrome)
    - [x] Linux (FF)
    - [x] Windows
    - [x] macOS
- [x] Update LICENSE.md
    - [x] Binomial Transcoder module
    - [x] KTX-Parse
- [x] Update CHANGES.md

cc @lilleyse - not sure if you wanted to take a closer look at this yet, especially with some of the to-do items left. thought it wouldn't hurt to open the PR to track progress.

cesium-concierge commented:
Thanks for the pull request @ebogo1!
* :heavy_check_mark: Signed CLA found.
* :grey_question: Changes to third party files were made.
    * Looks like a file in one of our ThirdParty folders (`ThirdParty/, Source/ThirdParty/`) has been added or modified. Please verify that it has a section in [LICENSE.md](https://github.com/CesiumGS/cesium/blob/master/LICENSE.md) and that its license information is up to date with this new version.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

ebogo1 commented:
As a note, I have been using Khronos's `ktx2ktx2` command-line tool for converting our KTX models to KTX2 - see  https://github.com/KhronosGroup/KTX-Software/blob/master/tools/ktx2ktx2/. The repo has instructions for building them with cmake. So far, `LogoETC1.ktx2` is the only file created with their tool, but I think it is more convenient than meshopt's gltfpack tool because it operates on .ktx files rather than .gltf/.glb.

ebogo1 commented on Source/Scene/KTX2Transcoder.js:
```diff
@@ -0,0 +1,89 @@
+  return KTX2Transcoder._readyPromise
+    .then(function (taskProcessor) {
+      var parameters;
+      if (ktx2Buffer instanceof ArrayBuffer) {
```
@lilleyse This is where checking for ArrayBuffers vs Uint8Arrays got a little messy

ebogo1 commented on Specs/Data/Models/Box-Textured-KTX2-Basis/CesiumTexturedBoxTest.gltf:
```diff
@@ -0,0 +1,210 @@
+{
```
This should be updated to be spec-compliant - see [glTF specs](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_texture_basisu).

ebogo1 commented on Specs/Scene/OctahedralProjectedCubeMapSpec.js:
```diff
@@ -14,7 +14,7 @@ describe(
 
     var environmentMapUrl =
-      "./Data/EnvironmentMap/kiara_6_afternoon_2k_ibl.ktx";
+      "./Data/EnvironmentMap/kiara_6_afternoon_2k_ibl.ktx2";
```
This seems to break the first two specs because the numbers don't match anymore. Will have to look into this; the aspect ratio of the new texture width and height in `"creates a packed texture with the right dimensions"` matches the old ratio, but the resolution is higher.

ebogo1 replied:
update: as of 11edcf9 the texture dimensions test passes but I had to bump the RGB epsilon from 5 to 6 in the correct projection spec.

ebogo1 commented:
> Investigate alternative transcoder module

There are two libraries here that I had in mind for this to-do item:
- [Basis transcoder](https://github.com/KhronosGroup/Basis-Universal-Transcoders) from Khronos
- [KTX-parse](https://github.com/donmccurdy/KTX-Parse) 

The latter might simplify the code in `parseKTX2.js` and seems like it fits with any of the available transcoder modules. But, it looks like the Khronos basis transcoder doesn't support transcoding from ETC1S whereas the Binomial transcoder supports both ETC1S and UASTC. The Khronos transcoder readme does state that there are plans to implement the missing formats.

@lilleyse - maybe it's a good idea to compare both the Binomial and Khronos transcoders' performance, and maybe use one over the other depending on format? Any chance we decide not to support a particular format in the first pass here?

ebogo1 commented on Source/Core/getUint64FromDataView.js:
```diff
@@ -23,10 +23,10 @@ function getUint64FromDataView(view, byteOffset, littleEndian) {
-    //9007199254740991) { // MAX_SAFE_INTEGER not available in IE11
-    console.warn(combined, "exceeds MAX_SAFE_INTEGER. Precision may be lost");
-  }
+  // if (!combined > Number.MAX_SAFE_INTEGER) {
```
this was commented out because of IE11 issues.

lilleyse commented on Specs/Scene/OctahedralProjectedCubeMapSpec.js:
```diff
@@ -14,7 +14,7 @@ describe(
     var octahedralMap;
```
Miscellaneous comments:

* Do a global search for `ktx`, `crn`, and `crunch` (with `Match Whole Word`). Remove all occurrences of `crn`/`crunch` (except in `CHANGES.md`), and see if `ktx` needs to be renamed to `ktx2` anywhere.
* Update `CHANGES.md` with a breaking changes section for 1.82 for `loadCRN` and `loadKTX` being removed
* I wonder if there's any way to build `Source/ThirdParty/basis_transcoder.wasm` so that it's smaller...

lilleyse replied:
@ebogo1 bump to finish doing the first bullet here

lilleyse replied:
I found some leftover occurrences: https://github.com/CesiumGS/cesium/pull/9513/commits/804f1ebdb695f62d46f3e8da1c74695dc1fb74e3

lilleyse commented on Source/Renderer/Context.js:
```diff
@@ -315,6 +316,7 @@ function Context(canvas, options) {
     "WEBKIT_WEBGL_compressed_texture_pvrtc",
   ]);
   this._etc1 = !!getExtension(gl, ["WEBGL_compressed_texture_etc1"]);
+  loadKTX2.setKTX2SupportedFormats(this._etc1, this._s3tc, this._pvrtc);
```
It's unfortunate that `loadKTX2` needs to be notified of the supported formats in this way but probably required since imagery layers don't have access to the context. I would just add a comment about that.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -1952,53 +1963,20 @@ function parseTextures(model, context, supportsWebP) {
+    } else if (
+      defined(texture.extensions) &&
+      defined(texture.extensions.KHR_texture_basisu) &&
+      FeatureDetection.supportsBasis
```
Related to the other comment, this should use `context.supportsBasis` since `Model` doesn't always have access to `scene`.

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -1925,21 +1926,31 @@ function imageLoad(model, textureId) {
+var ktx2Regex = /(^data:image\/ktx2)|(\.ktx2$)/i;
 
-function parseTextures(model, context, supportsWebP) {
+function parseTextures(model, context, supportsWebP, supportsBasis) {
```
`supportsBasis` is unused

lilleyse commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,335 @@
+import defined from "../Core/defined.js";
```
I'll hold off on reviewing this until you evaluate [KTX-parse](https://github.com/donmccurdy/KTX-Parse)

lilleyse commented on Specs/Renderer/TextureSpec.js:
```diff
@@ -480,18 +454,18 @@ describe(
 
-    it("draws the expected ETC1 compressed texture color", function () {
+    it("draws the expected KTX2 compressed texture color", function () {
       if (!context.etc1) {
         return;
       }
```
This early exit is not relevant either. I assume this code can be deleted because uncompressed KTX2 don't require the WebAssembly module to decode them.

lilleyse reviewed (commented):
Good start so far @ebogo1!

Similar to my review of https://github.com/CesiumGS/cesium/pull/9040 I mostly just left feedback on the files surrounding the KTX2 code. I'll take a pass at the transcoder / worker code after the smaller fixes are addressed.

ebogo1 commented:
4d0682a adds [KTX-Parse](https://github.com/donmccurdy/KTX-Parse) to Source/ThirdParty/Workers to cut down on the file size of `parseKTX2.js`. It also removed the need for `getUint64FromDataView.js`. Per @mramato's suggestion, I moved `ktx-parse.modern.js` from the node module's `dist` folder into our Workers folder so that it only got included from `parseKTX2.js`. I still have to update LICENSE.md.

ebogo1 commented on Apps/Sandcastle/gallery/Image-Based Lighting.html:
```diff
@@ -66,7 +66,7 @@
 
         var environmentMapURL =
-          "https://cesium.com/public/SandcastleSampleData/kiara_6_afternoon_2k_ibl.ktx";
+          "https://cesium.com/public/SandcastleSampleData/kiara_6_afternoon_2k_ibl.ktx2";
```
Yeah - wasn't sure because the regular .ktx is in master even though that one is also hosted externally.

lilleyse replied:
Yeah if nothing references it it can be removed

ebogo1 commented on Source/Core/VulkanConstants.js:
```diff
@@ -0,0 +1,287 @@
+/**
+ * Enum containing Vulkan Constant values by name.
+ *
+ * These match the constants from the [Vulkan 1.2]{@link https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#formats-definition}
```
Just to be sure, was the intended wording here "Vulkan 1.2 specification specifications."? 
```
...from the {@link https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#formats-definition|Vulkan 1.2 specification}.
```
seems better to me.

lilleyse replied:
Yup your wording is fine, I probably just forgot to select both lines when I made the suggested change.

ebogo1 commented on Source/Core/FeatureDetection.js:
```diff
@@ -304,6 +308,7 @@ var FeatureDetection = {
   supportsPointerEvents: supportsPointerEvents,
   supportsImageRenderingPixelated: supportsImageRenderingPixelated,
   supportsWebP: supportsWebP,
+  supportsBasis: supportsBasis,
```
I had to move the definition of `FeatureDetection.supportsBasis` a bit because of the circular dependency between `supportsBasis()` and `FeatureDetection.supportsWebAssembly()`.

ebogo1 commented:
@lilleyse Did a first pass over some of your feedback; haven't gotten to everything yet. `parseKTX2.js` should be good for review now that `KTX-Parse` is included.

ebogo1 commented on Apps/Sandcastle/gallery/Materials.html:
```diff
@@ -221,14 +221,7 @@
-          } else if (scene.getCompressedTextureFormatSupported("pvrtc")) {
-            compressedImageUrl = "../images/LogoPVR.ktx";
-          }
+          var compressedImageUrl = "../images/Cesium_Logo_Color.ktx2";
```
I got an ETC1S compressed .ktx2 working here in 8e708d1. It's currently upside-down though.

ebogo1 commented:
Since this turned into a deeper investigation than expected - 

> What tool did you use to convert the ktx to ktx2?

Initially I was using Khronos's `toktx` (and according to the .ktx2 metadata in #9040, so did Youssef). Testing a .png file compressed to .ktx2 with ETC1S encoding produced a texture that showed up as blank in CesiumJS. I tested this texture with a .glTF using the `KHR_texture_basisu` extension with no fallback (see [extension guidelines](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_texture_basisu)) in the [Babylon.js viewer](https://sandbox.babylonjs.com/) and it complained: `GL_INVALID_OPERATION: Invalid compressed image size.`. Surprisingly, the same .glTF I tested in Babylon worked in [Gestaltor](https://gestaltor.io/), which is recommended in Khronos's KTX2 [artist guide](https://github.com/KhronosGroup/3D-Formats-Guidelines/blob/main/KTXArtistGuide.md). 

Next I tried using [`gltfpack`](https://github.com/zeux/meshoptimizer/tree/master/gltf) to convert a regular glTF pointing to the original .png texture with the `-tc` flag to convert textures to .ktx2. Though this tool also uses `toktx` under the hood, this produced a compressed .glTF that worked in Babylon. CesiumJS in this PR now threw an error with the resulting .ktx2 about blank textures. 

I noticed the error only occurred when all mip levels of the .ktx2 were returned from `parseKTX2`, and found that `Material.js` didn't appear to support mip levels the same way that `Model.js` did. I tweaked the logic in `parseKTX2` and in `Material` and gltfpack's .ktx2 (CesiumLogoFlat.ktx2) now works in the Sandcastle and is down from 1.1 MB to 5.9 kB. But, now it's upside-down.. still looking into this bit.

Next is trying to update the original environment map to .ktx2 using gltfpack. I'm still unsure where I'm using `toktx` incorrectly but maybe digging into the source code of gltfpack can give me an idea. There should probably be a spec to test for Materials with multiple mip levels.

ebogo1 commented:
> I'm still unsure where I'm using toktx incorrectly

My mistake seems to be in trying to generate images with resolutions that are not multiples of 4.

ebogo1 commented:
da30ff7 Changes the logic in `parseKTX2` to allow UASTC files to make it to the transcoder. This seems to work fine even though we aren't setting the recommended pixel formats (see https://github.com/CesiumGS/cesium/pull/9513#discussion_r627682799). I added a UASTC ktx2 file to the Materials Sandcastle but it's quite larger than the ETC1S one (~65kB). I could try to compress it to a worse UASTC quality, but I thought it would be nice to show that both formats work in the Sandcastle. 

Next is figuring out how to tie in the `"KTX2 pixel format"` case to the transcoder because I believe the module supports it.

lilleyse commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,238 @@
+  var internalFormat, transcoderFormat;
+  var BasisFormat = transcoderModule.transcoder_texture_format;
+  if (supportedTargetFormats.astc && !hasAlpha) {
+    internalFormat = PixelFormat.RGB_ETC1;
```
Yeah add whatever constants are needed in `PixelFormat` and `WebGLConstants`. The constants are defined here:

https://developer.mozilla.org/en-US/docs/Web/API/EXT_texture_compression_bptc
https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_compressed_texture_astc

You'll need to add support for `EXT_texture_compression_bptc` in `Context` as well.

lilleyse replied:
Actually you might need to go the specs themselves to get the constants:

https://www.khronos.org/registry/webgl/extensions/EXT_texture_compression_bptc/
https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_astc/

ebogo1 commented:
> Investigate alternative transcoder module

Looks like [Babylon uses the smaller transcoder](https://doc.babylonjs.com/divingDeeper/materials/using/ktx2Compression) when possible and falls back to the Binomial one when needed. Three.js only uses the Binomial one (see [here](https://github.com/mrdoob/three.js/tree/dev/examples/js/libs/basis)). 

cc @lilleyse

ebogo1 commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -1,7 +1,7 @@
 import { loadKTX2 } from "../../Source/Cesium.js";
+import { KTX2Transcoder } from "../../Source/Cesium.js";
```
I think there should be a `KTX2TranscoderSpec` testing the parsing code, while `loadKTX2Spec` focuses on error/undefined handling?

ebogo1 commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -405,247 +251,13 @@ describe("Core/loadKTX2", function () {
     }).toThrowDeveloperError();
   });
 
-  it("creates and sends request without any custom headers", function () {
```
I couldn't find other examples where we have tests like this.. is it important to test url request behavior from `loadKTX2`?

ebogo1 commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -664,198 +276,90 @@ describe("Core/loadKTX2", function () {
+    var resource = Resource.createIfNeeded("./Data/Images/Green4x4.ktx2");
+    var loadPromise = resource.fetchArrayBuffer();
+    return loadPromise.then(function (buffer) {
+      var promise = KTX2Transcoder.transcode(buffer, { pvrtc: true });
```
The `{ pvrtc: true }` object is kinda phony, but at least one format needs to be "supported" for the logic in `parseKTX2` to go forward without error-ing. Is there a better way to get the formats here? Should all formats be tested in specs somehow?

ebogo1 commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -664,198 +276,90 @@ describe("Core/loadKTX2", function () {
-      }
+  it("returns a promise that resolves to a compressed texture", function () {
+    var resource = Resource.createIfNeeded(
+      "./Data/Images/Green4x4Compressed.ktx2"
```
Currently this points to an ETC1S ktx2 file I made. Should there be UASTC versions of the regular + mipmap file?

ebogo1 commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -664,198 +276,90 @@ describe("Core/loadKTX2", function () {
+    });
   });
 
   it("returns a promise that resolves to a compressed texture containing the all mip levels of the original texture", function () {
```
this doesn't pass yet, still have to add support for compressed texture mipmaps

ebogo1 reviewed (commented):
@lilleyse A few questions about these changes -

ebogo1 commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,241 @@
+    header.vkFormat === VulkanConstants.VK_FORMAT_B10G11R11_UFLOAT_PACK32
+      ? WebGLConstants.R11F_G11F_B10F
+      : internalFormat;
+  internalFormat =
```
@lilleyse Since the math in the cubemap shader has to be different for `..._UNORM` formats, what would be a good way to catch that case here? Should this be an `if (header.vkFormat === VulkanConstants.R8G8B8A8_UNORM || header.vkFormat === VulkanConstants.R16G16B16A16_UNORM ...)` statement?

ebogo1 reviewed (commented):
After the last two commits `Material` should work with KTX2 containing embedded mipmaps. Still need to fix the spec in `loadKTX2` for compressed files with mipmap levels.

The logic around compressed textures and mipmaps in `Model` still needs to be worked.

ebogo1 commented on Source/Core/PixelFormat.js:
```diff
@@ -47,6 +47,14 @@ var PixelFormat = {
+   * @type {Number}
+   * @constant
+   */
+  RGBA_UNORM: WebGLConstants.UNSIGNED_NORMALIZED,
```
@lilleyse I'm pretty certain this is the wrong WebGL constant to use, but looking at the docs [here](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Constants) I couldn't find one that matches the 8-bit UNORM format of the IBL sampler... SNORM is the closest I see. Any thoughts here?

lilleyse replied:
`RGBA` is implicitly UNORM when paired with `PixelDatatype.UNSIGNED_BYTE` so an additional pixel format does not have to be added

lilleyse replied:
`RGBA` is implicitly UNORM when paired with `PixelDatatype.UNSIGNED_BYTE` so an additional pixel format does not have to be added

ebogo1 commented on Source/Scene/OctahedralProjectedCubeMap.js:
```diff
@@ -302,11 +303,22 @@ OctahedralProjectedCubeMap.prototype.update = function (frameState) {
   }
 
+  var defines = [];
+  if (
```
Still to be tested with the SFLOAT formats from the IBL sampler but I imagined that the new `isNormalizedFormat` function would catch both when pushing the define.

ebogo1 reviewed (commented):
A few comments about the last couple commits to keep track of the changes. I reworked OctahedralProjectedCubeMap to use a shader define for normalized vs unnormalized formats.

ebogo1 commented:
@lilleyse Should be ready for another look since most of the logic should be there now. The last commit cleaned up `parseKTX2` and added checks for 16 and 32-bit SFLOAT formats. 

I also now made changes to store a `PixelDatatype` in `CompressedTextureBuffer` so that `OctahedralProjetedCubeMap` knows whether or not pixel values are normalized. I expect this will also be useful because there needs to be changes in some part of the code for non 8-bit data to work with cubemaps. 

Other than that, the main items left are to add compressed texture mipmap support in `Model.js` and to add the second transcoder module. Neither should impact the current code changes by much.

Here are the IBL cubemaps I've been testing with (generated by the [Khronos sampler](https://github.com/KhronosGroup/glTF-IBL-Sampler)) and a bonus ETC1S compressed texture with a mip pyramid - 
[ktx2-testing.zip](https://github.com/CesiumGS/cesium/files/6601331/ktx2-testing.zip)

mramato commented on Source/Scene/Material.js:
```diff
@@ -878,13 +888,12 @@ function createTexture2DUpdateFunction(uniformId) {
           promise = resource.fetchImage();
         }
+
         when(promise, function (image) {
```
I think this code was here already, but we should never call `when(promise` like this, we should always use `promise.then`.  Also, it appears that there is no error handling here, which is weird.  Can you write up a clean-up issue for us to review promises in Material.js

ebogo1 replied:
opened https://github.com/CesiumGS/cesium/issues/9610

mramato commented on Source/Scene/Material.js:
```diff
@@ -878,13 +888,12 @@ function createTexture2DUpdateFunction(uniformId) {
-          promise = loadKTX(resource);
-        } else if (crnRegex.test(resource.url)) {
-          promise = loadCRN(resource);
+        if (ktx2Regex.test(resource.url)) {
```
I guess this was already like this, but it means that we require KTX files to have a specific extension, which is generally bad practices (especially on the web) normally we would check the header magic or similar.

ebogo1 replied:
@lilleyse We discussed this offline - should I open an issue for this?

lilleyse replied:
Yeah open an issue

ebogo1 replied:
opened https://github.com/CesiumGS/cesium/issues/9640

mramato commented on Source/Scene/Material.js:
```diff
@@ -435,6 +434,17 @@ Material.prototype.update = function (context) {
+    var mipLevels;
+    if (Array.isArray(image)) {
+      // highest detail mip should be level 0
+      mipLevels = image.slice(1, image.length).map(function (mipLevel) {
```
Throughout, use `arraySlice` instead of `.slice`

mramato replied:
In fact, from a performance standpoint, we wouldn't actually use `map` and `slice` here at all.

lilleyse replied:
This area shouldn't be a performance problem since it only gets called once when the texture is first loaded

ebogo1 commented:
I expected ab6c6cf to make 16- and 32-bit textures to work with cubemap, but it looks like only 16-bit works for now. The fix was to pass the `PixelDatatype` to `CubeMap`, which otherwise assumes 8-bit. I'll look at this some more; since it was missing from the above .zip here's the 32-bit cubemap I'm testing with:

[iblKiara32.ktx2.zip](https://github.com/CesiumGS/cesium/files/6641079/iblKiara32.ktx2.zip)

lilleyse commented on Source/Scene/OctahedralProjectedCubeMap.js:
```diff
@@ -302,11 +303,28 @@ OctahedralProjectedCubeMap.prototype.update = function (frameState) {
   }
 
+  var defines = [];
+  // Datatype is defined if it is a normalized type (i.e. ..._UNORM, ..._SFLOAT)
```
Is this the right fix?

```suggestion
  // Datatype is defined if it is a normalized type (i.e. ..._UNORM)
```

lilleyse replied:
Maybe we could discuss this and the comment below offline because I'm a little confused about when `pixelDatatype` is defined and how the behavior differs for unorm vs. float textures.

ebogo1 replied:
Both UNORM and SFLOAT outputs from the Khronos IBL sampler output rgb values in [0, 1], so I push the normalized "define" in both cases. In the shader, the behavior should be the same for UNORM and SFLOAT but it might be possible to get a cubemap with unnormalized rgb (i.e. Filament sampler).

lilleyse replied:
> Both UNORM and SFLOAT outputs from the Khronos IBL sampler output rgb values in [0, 1]

Is this true? I would've thought HDR float textures would go beyond the [0, 1] range. The shader is fine though, the "normalized" naming just trips me up.

lilleyse commented on Source/Shaders/OctahedralProjectionFS.glsl:
```diff
@@ -3,8 +3,12 @@ uniform samplerCube cubeMap;
+    vec4 rgba = textureCube(cubeMap, v_cubeMapCoordinates);
+    #ifdef RGBA_NORMALIZED
+        gl_FragColor = vec4(rgba.rgb, 1.0);
+    #else
+        float m = rgba.a * 16.0;
+        vec3 r = rgba.rgb * m;
+        gl_FragColor = vec4(r * r, 1.0);
+    #endif
```
This still looks like the RGBM encoding. Does anything produced by https://github.com/KhronosGroup/glTF-IBL-Sampler go down this path?

ebogo1 replied:
nothing from the Khronos sampler reaches here - more info in above comment

lilleyse replied:
Got it, thanks!

lilleyse commented on Source/Scene/KTX2Transcoder.js:
```diff
@@ -0,0 +1,90 @@
+    })
+    .then(function (result) {
+      var levelsLength = result.length;
+      var faceKeys = Object.keys(result[0]);
```
Is `Object.keys` required? Is `result` an array?

ebogo1 replied:
`result` is an array of objects, so this returns an array of the `positiveX ... negativeZ` face keys.

lilleyse commented on Apps/Sandcastle/gallery/Image-Based Lighting.html:
```diff
@@ -59,14 +59,20 @@
         //Sandcastle_Begin
```
I see a slight difference between the environment map in this branch and master. Is the image rotated differently? Normally this wouldn't really matter but the spherical harmonics are no longer aligned with the environment map.

![ibl_difference](https://user-images.githubusercontent.com/915398/122145710-c0dfc600-ce23-11eb-8b53-8f0e9f64183e.jpg)

ebogo1 replied:
not sure what a good workaround here is.. the Khronos tool does not generate spherical harmonics.

ebogo1 replied:
doesn't look like Filament's spherical harmonics match the Khronos sampler's output - 
![image](https://user-images.githubusercontent.com/31491650/122988364-a9e32b80-d36f-11eb-9a1f-8eb6c04029f3.png)

there are also weird patterns at certain angles both without spherical harmonics and with the misaligned ones - 
![image](https://user-images.githubusercontent.com/31491650/122989024-58876c00-d370-11eb-9d14-4b1ffa98cb0d.png)

lilleyse replied:
The spherical harmonic problem was fixed in https://github.com/CesiumGS/cesium/pull/9513/commits/e6b833d6be9e5fc12df97614a86063e060b788dd by updating the shader code to match to the newer filament code.

lilleyse commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+import RuntimeError from "../Core/RuntimeError.js";
+import VulkanConstants from "../Core/VulkanConstants.js";
+import PixelDatatype from "../Renderer/PixelDatatype.js";
+import { read } from "../ThirdParty/Workers/ktx-parse.modern.js";
```
Is `ktx-parse.modern.js` ok to use? Is it compatible with IE11? The reason I ask is because KTX-Parse can also generate `ktx-parse.js` which looks more es5-like, though doesn't seem to work out of the box like `ktx-parse.modern.js`

ebogo1 replied:
I'm not sure; the [docs](https://github.com/donmccurdy/ktx-parse#api-documentation) don't seem to mention supported browsers. Maybe cc @mramato here?

Did we want to support KTX2 in IE11?

lilleyse replied:
No action needed since we plan to drop support for IE11 soon. Which reminded me, we should add `ktx-parse` to the list of npm packages: https://github.com/CesiumGS/cesium/issues/9473#issuecomment-866257585

lilleyse commented on Apps/Sandcastle/gallery/Image-Based Lighting.html:
```diff
@@ -59,14 +59,20 @@
 
         var environmentMapURL =
-          "https://cesium.com/public/SandcastleSampleData/kiara_6_afternoon_2k_ibl.ktx";
+          "https://cesium.com/public/SandcastleSampleData/kiara_6_afternoon_2k_ibl.ktx2";
```
Make sure to update this file once everything is final

lilleyse replied:
Do you have all versions of kiara to test with (8-bit unorm, 16-bit float, 32-bit float)? I'm curious to try them all out.

lilleyse commented on Apps/Sandcastle/gallery/Materials.html:
```diff
@@ -218,17 +218,25 @@
+        function applyUASTCCompressedTextureMaterial(primitive, scene) {
+          Sandcastle.declare(applyUASTCCompressedTextureMaterial); // For highlighting in Sandcastle.
+
+          var compressedImageUrl = "../images/Cesium_Logo_UASTC.ktx2";
```
`Apps/Sandcastle/images/Cesium_Logo_Flat.ktx2` can be deleted as it is no longer being used in the materials sandcastle

lilleyse commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+import arraySlice from "../Core/arraySlice.js";
```
`parseKTX2` and `KTX2Transcoder` belong in `Core` since they are called from `loadKTX2` which is in `Core`

lilleyse commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+
+// Flags
+var colorModelETC1S = 163;
+var colorModelUASTC = 166;
```
From https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_texture_basisu#ktx-v2-images-with-basis-universal-supercompression

> To cover a broad range of use cases, this extension allows two Basis Universal formats: ETC1S with BasisLZ and UASTC with optional Zstandard compression.

Have you tried loading UASTC with Zstandard compression? Does the basis transcoder handle that?

lilleyse reviewed (commented):
@ebogo1 the recent changes look good and I'm starting to feel like the PR is stabilizing. I didn't review the tests because they are still in flux.

lilleyse commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+  // Determine target format based on platform support
+  var internalFormat, transcoderFormat;
+  var BasisFormat = transcoderModule.transcoder_texture_format;
+  if (supportedTargetFormats.astc && hasAlpha) {
+    internalFormat = PixelFormat.RGBA_ASTC;
+    transcoderFormat = BasisFormat.cTFASTC_4x4_RGBA;
+  } else if (supportedTargetFormats.bc7) {
+    internalFormat = PixelFormat.RGBA_BC7;
+    transcoderFormat = BasisFormat.cTFBC7_RGBA;
```
Couple questions about the precedence of GPU compressed texture formats

* What's the advantage of BC7 over DXT5? Is BC7 higher quality? Is it better for normal maps?
* This order makes more sense for UASTC, but should ETC1S transcode to ETC1/DXT1 by default? They are smaller in GPU memory than ASTC/BC7 and might be good enough quality.

ebogo1 replied:
- BC7 is higher quality than the older BC formats; I found [this post](https://www.reedbeta.com/blog/understanding-bcn-texture-compression-formats/#bc1) comparing them:
![image](https://user-images.githubusercontent.com/31491650/122684259-6a330d00-d1d2-11eb-9e45-bfd8d647995e.png)
![image](https://user-images.githubusercontent.com/31491650/122684254-5edfe180-d1d2-11eb-8e72-b7ea0317e4c7.png)
Though based on their example of a full res texture maybe DXT5 is sufficient for CesiumJS?
![image](https://user-images.githubusercontent.com/31491650/122684299-a1092300-d1d2-11eb-885f-a46f8b2a05fb.png)
![image](https://user-images.githubusercontent.com/31491650/122684331-cb5ae080-d1d2-11eb-9470-df0c3097bfc1.png)
- According to [these comments](https://github.com/BinomialLLC/basis_universal/blob/c4c0db735a6a799702cd4e7309df64b4f7f59e39/webgl/ktx2_encode_test/index.html#L312) in the Binomial WebGL encoding example, the preferred format for ETC1S without alpha is ETC1. Not sure how they decided that but the format precedence in their code doesn't seem to force ETC1 for all ETC1S files either... I think the idea is that ETC1 does not support alpha, so it should not be used unless the ktx2 is in ETC1S + no alpha. Otherwise a format with an alpha channel is needed like BC7 (according to the above post and some other sources DXT1 only supports a 1-bit alpha channel?).

ebogo1 commented on Source/Core/FeatureDetection.js:
```diff
@@ -308,6 +308,12 @@ var FeatureDetection = {
+  return FeatureDetection.supportsWebAssembly() && scene.context.supportsBasis;
+}
+
+FeatureDetection.supportsBasis = supportsBasis;
```
What's the right way to add docs here? It thinks `supportsBasis` is a member of `FeatureDetection` rather than a method. Maybe something like this?
```js
FeatureDetection.supportsBasis = function (scene) {
    return supportsBasis(scene);
}
```

lilleyse replied:
Actually, why can't it be:

```js
FeatureDetection.supportsBasis = function (scene) {
  return FeatureDetection.supportsWebAssembly() && scene.context.supportsBasis;
}
```

ebogo1 commented on Source/Core/loadKTX2.js:
```diff
@@ -0,0 +1,114 @@
+ *     <li>3D textures</li>
+ *     <li>Texture Arrays</li>
+ *     <li>Video</li>
+ *     <li>Compressed Mipmaps</li>
```
they are supported :+1:

ebogo1 commented on Source/Core/loadKTX2.js:
```diff
@@ -0,0 +1,114 @@
+ * </ul>
+ * </p>
+ *
+ * @exports loadKTX2
```
cc @mramato?

lilleyse replied:
I was looking at the older `loadKTX` for what it's worth

lilleyse replied:
I stumbled upon an older conversation on slack that explains this better from @mramato 

> In general, avoid using @exports, it was a bad habit that was 99% removed form the Cesium code base when we moved to TS.

https://cesium.slack.com/archives/CKDFTTDCN/p1608082515024700

ebogo1 commented on Source/Renderer/Context.js:
```diff
@@ -741,6 +754,19 @@ Object.defineProperties(Context.prototype, {
+   * access to ASTC compressed textures.
+   * @memberof Context.prototype
+   * @type {Boolean}
+   * @see {@link https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_etc1/}
```
what's wrong here? The link works for me.

lilleyse replied:
It should be https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_astc/

ebogo1 commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+ */
+function parseKTX2(data, supportedTargetFormats, transcoderModule) {
+  if (!defined(supportedTargetFormats)) {
+    throw new RuntimeError(
+      "KTX2 Loader does not know which GPU formats are supported."
+    );
+  }
```
I added `Check.typeOf.object` for supportedTargetFormats and transcoderModule; like you mentioned in another comment it'd be impossible for `data` to be undefined because of the early return in `loadKTX2`.

ebogo1 commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+
+    for (var j = 0; j < header.faceCount; ++j) {
+      var faceLength =
+        header.typeSize *
```
From the [KTX2 spec](https://github.khronos.org/KTX-Specification/), "`typeSize` specifies the size of the data type in bytes used to upload the data to a graphics API". I've added a comment 👍

ebogo1 commented on Source/Scene/Material.js:
```diff
@@ -435,6 +434,17 @@ Material.prototype.update = function (context) {
     var image = loadedImage.image;
 
+    // Images transcoded from KTX2 can contain multiple mip levels:
+    // https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_texture_basisu
```
a link to KTX2 spec might be better? [https://github.khronos.org/KTX-Specification/](https://github.khronos.org/KTX-Specification/)

lilleyse replied:
Sure, is there a specific section that you can link?

ebogo1 replied:
[https://github.khronos.org/KTX-Specification/#_mip_level_array](https://github.khronos.org/KTX-Specification/#_mip_level_array) seems good to me.

lilleyse replied:
Ah, the link still needs to be changed from https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_texture_basisu to https://github.khronos.org/KTX-Specification/#_mip_level_array

ebogo1 commented on Source/Scene/Material.js:
```diff
@@ -435,6 +434,17 @@ Material.prototype.update = function (context) {
+    }
+
     var sampler = new Sampler({
       minificationFilter: this._minificationFilter,
       magnificationFilter: this._magnificationFilter,
```
I wasn't sure whether it'd be on us to throw an error or warning in this case. Different browsers handle this differently.

lilleyse replied:
Actually this isn't a concern because `Material` doesn't call `generateMipmap`

ebogo1 commented:
@lilleyse I did a first pass of changes in response to your comments but I'll ping you again once everything's addressed. 

Here's a [Google Drive folder](https://drive.google.com/drive/folders/1WbhyutrHxn9CgyvavNHpXrwGtiwr13pB?usp=sharing) with most of the KTX2 sample data I've been testing with.

ebogo1 commented on Specs/Renderer/TextureSpec.js:
```diff
@@ -480,18 +454,18 @@ describe(
     });
 
-    it("draws the expected ETC1 compressed texture color", function () {
+    it("draws the expected KTX2 compressed texture color", function () {
```
it is uncompressed 👍

ebogo1 commented on Source/Core/loadKTX2.js:
```diff
@@ -0,0 +1,114 @@
+      var copy = arraySlice(data);
+      return KTX2Transcoder.transcode(copy, supportedTranscoderFormats);
+    }
+    return when.reject(new RuntimeError("failed to load KTX2 data."));
```
is it possible for a Resource loadPromise pointing to an online URL to eventually resolve to an undefined buffer?

lilleyse replied:
I don't think so, I would expect the promise to be rejected instead

ebogo1 replied:
Should this line just be deleted then?

lilleyse replied:
Yeah, and the if statement can be removed

ebogo1 commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+  } else if (supportedTargetFormats.bc7) {
+    internalFormat = PixelFormat.RGBA_BC7;
+    transcoderFormat = BasisFormat.cTFBC7_RGBA;
+  } else if (supportedTargetFormats.etc1 && !hasAlpha) {
```
From [Wikipedia](https://en.wikipedia.org/wiki/Ericsson_Texture_Compression), there is no support for alpha in ETC1.

ebogo1 commented on Source/Scene/Model.js:
```diff
@@ -2891,6 +2893,7 @@ function createTexture(gltfTexture, model, context) {
       context: context,
       source: {
         arrayBufferView: gltfTexture.bufferView,
+        mipLevels: gltfTexture.mipLevels,
```
the `mipLevels` argument below is never used

ebogo1 commented on Source/Scene/parseKTX2.js:
```diff
@@ -0,0 +1,225 @@
+
+  for (var i = 0; i < header.levels.length; ++i) {
+    var level = (result[i] = {});
+    var levelBuffer = arraySlice(header.levels[i].levelData);
```
My thinking was based on this part in the [KTX-Parse docs](https://github.com/donmccurdy/ktx-parse#read) - 
> The container's mip levels and other binary data are pointers into the original file, not copies, so the original file should not be overwritten after reading.

I thought it'd be safer to not touch the original data when it gets passed to the `result` object,  but there might be a better way... the IBL sandcastle gets messed up without arraySlice here.

lilleyse replied:
Does our code actually modify the data? Also is it just the cube map that gets corrupted or does it happen to regular images as well?

On a related topic, whether it's possible to remove `arraySlice` or not, the underlying `ArrayBuffer` should still be added to `transferrableObjects` so that it's transferred rather than copied when it goes back to the main thread.

This can be done by adding `transferableObjects` to the `transcode` function in `transcodeKTX2` and passing it along to `parseKTX2`. `parseKTX2` should then push `ArrayBuffer`'s that are used by the result.

In general CesiumJS doesn't take advantage of transferring data back to the main thread as much as it should, though we do the reverse all the time.

```
function transcode(parameters, transferableObjects) {
  var ktx2Buffer = parameters.ktx2Buffer;
  var supportedTargetFormats = parameters.supportedTargetFormats;
  return parseKTX2(ktx2Buffer, supportedTargetFormats, transcoderModule, transferableObjects);
}
```

lilleyse replied:
Back to `arraySlice`...

I don't think we need it, the typed array constructors just needs to be given the proper byte offset into the `ArrayBuffer`.

`arraySlice` was hiding the problem because it created a new `ArrayBuffer` copy of the view. 

```js
      var faceByteOffset = levelBuffer.byteOffset + faceLength * j;

      var faceView;
      if (!defined(datatype) || PixelDatatype.sizeInBytes(datatype) === 1) {
        faceView = new Uint8Array(levelBuffer.buffer, faceByteOffset, faceLength);
      } else if (PixelDatatype.sizeInBytes(datatype) === 2) {
        faceView = new Uint16Array(levelBuffer.buffer, faceByteOffset, faceLength);
      } else {
        faceView = new Float32Array(levelBuffer.buffer, faceByteOffset, faceLength);
      }
```

lilleyse replied:
With both of these changes I think we can get to zero copies!

ebogo1 replied:
I think the above fix only works for Uint8Array because `faceLength` should be the same in all three cases regardless of `header.typeSize`. With a couple small tweaks I think this is working now - 9273229 has no `arraySlice` in the worker and works for all three IBLs in the ktx2 sample folder.

Did we reach a conclusion about needing `arraySlice` in the main thread to avoid accessing the same glb for ktx2 + vertex data?

lilleyse replied:
> Did we reach a conclusion about needing `arraySlice` in the main thread to avoid accessing the same glb for ktx2 + vertex data?

Yes - I pushed the change in this commit: https://github.com/CesiumGS/cesium/pull/9513/commits/b33dc12d6b787fabf1034658907e7183e828e591. However I used the `Uint8Array` instead of `arraySlice` for performance reasons and because I knew the input and output would always be `Uint8Array`.

ebogo1 commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -0,0 +1,194 @@
+import { Resource } from "../../Source/Cesium.js";
+import { RuntimeError } from "../../Source/Cesium.js";
+
+describe("Core/loadKTX2", function () {
```
@lilleyse These should be good for review now. I cleaned up some tests and removed a few others:
- Removed "mipmaps are supported" test because `OctahedralProjectedCubeMapSpec` confirms this
- Removed a couple tests checking for errors when the .ktx2 had mismatching type/typeSize with compressed textures. I could add these back, but since compressed ktx2 buffers don't always fit evenly into Uint32Arrays it's require more precise byte changes to manipulate the header info. 

Do we want to throw specific errors when header magic fields like `vkFormat` etc. are invalid? I expect these to be handled as "Invalid KTX2 file." errors through the transcoder module.

lilleyse replied:
The catch-all "Invalid KTX2 file" is ok for now. If it becomes a problem we can make the errors more granular.

ebogo1 commented on Source/Core/loadKTX2.js:
```diff
@@ -0,0 +1,114 @@
+ * @exception {RuntimeError} The base internal format must be the same as the format for uncompressed textures.
+ * @exception {RuntimeError} 3D textures are not supported.
+ * @exception {RuntimeError} Texture arrays are not supported.
+ * @exception {RuntimeError} Cubemaps are not supported.
```
they are now - between these and the specs all the possible error messages are covered. I'm not sure what would cause the `startTranscoding()` and `transcodeImage()` failures, but the Binomial example has those in their [WebGL demo]https://github.com/BinomialLLC/basis_universal/blob/c4c0db735a6a799702cd4e7309df64b4f7f59e39/webgl/ktx2_encode_test/index.html#L393)..

ebogo1 replied:
Like I mentioned in my review comment on loadKTX2Spec, I'm not sure what level or granularity we want with these errors.

lilleyse replied:
This level of granularity is fine

ebogo1 commented on Source/Core/loadKTX2.js:
```diff
@@ -0,0 +1,114 @@
+    if (defined(data)) {
+      // data array can be a view of a shared buffer,
+      // so make a copy that can be given to the worker.
+      var copy = arraySlice(data);
```
I think it's safe to remove `arraySlice` here because the ktx2 buffers are always coming from files that are loaded just for `loadKTX2` - testing without arraySlice here works fine in Sandcastle and specs (which all call loadKTX2 with a resource or url). Would there ever be a case where the data passed to `loadKTX2` would be needed somewhere else too?

lilleyse replied:
A KTX2 could be embedded in a glb. Does that still work? I wouldn't expect it to be a problem.

ebogo1 replied:
Just tested this with a .glb and it works fine. I added the .glb to the ktx2 sample data folder just in case.

ebogo1 commented:
@lilleyse db76034 changes the parseKTX2 format precedence. It should be in line with both three.js and Babylon other than its preference of smaller formats over BC7 for ETC1S. Also, I haven't found any specs online that state ETC1S is incompatible with ASTC but I haven't tested it yet. I think the best way to check this is to push a dummy commit that forces ASTC fallback on ETC1S ktx2s, then run the deployment on my phone (neither of my laptops support the ASTC extension).

I also added preference for ETC2 over ETC1 whenever possible. There are separate ETC2 WebGL constants for non-alpha (`RGB8_ETC2`) and alpha (`RGBA8_ETC2_EAC`). Both three and Babylon use the ETC1 transcode target if the ktx2 does not have alpha as the transcoder module does not have an RGB ETC2 format. I tested the db76034 deployment on my phone and it looks like this works fine; this is expected because ETC2 is [backwards compatible (see slide 10)](https://www.khronos.org/assets/uploads/developers/library/2012-siggraph-opengl-es-bof/Ericsson-ETC2-SIGGRAPH_Aug12.pdf). Might be worth doing more rigorous testing on mobile with alpha vs non alpha ktx2s?

The last difference between the new logic and Babylon's is that we don't fall back to RGBA32 as a last resort. The existing `PixelFormat.RGBA` does not work with the transcoder's `cTFRGBA_32` format, probably because it doesn't know that the texture is floating point. Any thoughts on how to handle this?

lilleyse commented on CHANGES.md:
```diff
@@ -2,12 +2,21 @@
+  - Added support for glTF models with the [`KHR_texture_basisu`](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_texture_basisu/README.md) extension.
+  - Added support for 8-bit, 16-bit float, and 32-bit float KTX2 specular environment maps.
+  - Added support for KTX2 images in `Material`.
+  - Added new `PixelFormat` and `WebGLConstants` enums from WebGL extensions `WEBGL_compressed_texture_astc` and `EXT_texture_compression_bptc`.
```
Add `WEBG_compressed_texture_etc` too

lilleyse commented on Source/Core/FeatureDetection.js:
```diff
@@ -318,6 +318,16 @@ var FeatureDetection = {
+ * Detects whether the current browser supports Basis Universal textures and the web assembly modules needed to transcode them.
+ *
+ * @param {Scene} scene
+ * @returns {Boolean} true if the browser supports web assembly modules and the scene supports Basis Universal textures, false if not.
```
Note - if we support the RGBA fallback we won't need `scene.context.supportsBasis` anymore.

lilleyse commented on Source/Renderer/Texture.js:
```diff
@@ -149,10 +149,18 @@ function Texture(options) {
       throw new DeveloperError(
         "When options.pixelFormat is PVRTC compressed, this WebGL implementation must support the WEBGL_texture_compression_pvrtc extension. Check context.pvrtc."
       );
+    } else if (PixelFormat.isASTCFormat(internalFormat) && !context.astc) {
+      throw new DeveloperError(
+        "When options.pixelFormat is ASTC compressed, this WebGL implementation must support the WEBGL_texture_compression_astc extension. Check context.astc."
+      );
     } else if (PixelFormat.isETC1Format(internalFormat) && !context.etc1) {
       throw new DeveloperError(
         "When options.pixelFormat is ETC1 compressed, this WebGL implementation must support the WEBGL_texture_compression_etc1 extension. Check context.etc1."
       );
+    } else if (PixelFormat.isBC7Format(internalFormat) && !context.bc7) {
+      throw new DeveloperError(
+        "When options.pixelFormat is BC7 compressed, this WebGL implementation must support the EXT_texture_compression_bptc extension. Check context.bc7."
+      );
```
A similar check should be added for `etc`

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -1674,7 +1674,11 @@ Scene.prototype.getCompressedTextureFormatSupported = function (format) {
       context.pvrtc) ||
     ((format === "WEBGL_compressed_texture_etc1" || format === "etc1") &&
-      context.etc1)
+      context.etc1) ||
+    ((format === "WEBGL_compressed_texture_astc" || format === "astc") &&
+      context.astc) ||
+    ((format === "EXT_texture_compression_bptc" || format === "bc7") &&
+      context.bc7)
```
Also add `etc`

lilleyse reviewed (commented):
Still need to get to tests, but a few comments for the main code.

lilleyse commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -0,0 +1,322 @@
+  it("throws with unknown supported formats", function () {
+    var promise = KTX2Transcoder.transcode(new Uint8Array());
+    expect(promise).toBeDefined();
+    return promise.then(
+      function () {
+        return;
+      },
+      function (error) {
+        expect(function () {
+          throw error;
+        }).toThrowDeveloperError();
+      }
+    );
+  });
```
It would be better if `KTX2Transcoder.transcode` checked that the parameters are defined instead of checking them in the worker. Generally it's good to fail fast like that. It should also simplify the test.

Also the preferred way to handle promises is to chain `.then` and `.otherwise` like below. Also `return` should be `fail()` so that jasmine fails if the promise resolves. Make sure to update other tests in this file.

```
    return promise
      .then(function (value) {
        return; // <------- this should actually be fail() so that jasmine fails if the promise somehow resolves
      })
      .otherwise(function (error) {
        expect(function () {
          throw error;
        }).toThrowDeveloperError();
      });
```

lilleyse commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -0,0 +1,322 @@
+    });
+  });
+
+  it("returns a promise that resolves to a compressed texture containing the all mip levels of the original texture", function () {
```
```suggestion
  it("returns a promise that resolves to a compressed texture containing all mip levels of the original texture", function () {
```

lilleyse commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -0,0 +1,322 @@
+    var resolvedValue;
+    var rejectedError;
+    var promise = loadKTX2(invalidKTX.buffer);
+    expect(promise).toBeDefined();
```
It's ok not to do `expect(promise).toBeDefined();` in these tests. The test will fail if it's undefined.

lilleyse commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -0,0 +1,322 @@
+    expect(promise).toBeDefined();
+    return promise.then(
+      function (value) {
+        resolvedValue = value;
```
Can this be `fail()` instead?

lilleyse commented on Specs/Data/Models/Box-Textured-KTX2-Mipmap/CheckerboardTexturedBoxTest.gltf:
```diff
@@ -0,0 +1,206 @@
+      "mimeType": "image/ktx2"
+    }
+  ],
+  "samplers": [],
```
This file fails validation https://github.khronos.org/glTF-Validator/

I think because empty arrays for top level properties are disallowed

ebogo1 replied:
I removed the samplers array - doesn't look the validator checks for KHR_texture_basisu but not having a sampler with a mipmap doesn't seem to break any [rules](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_texture_basisu#additional-requirements).

lilleyse commented on Specs/Renderer/TextureSpec.js:
```diff
@@ -1,6 +1,7 @@
 import { Color } from "../../Source/Cesium.js";
-import { loadKTX } from "../../Source/Cesium.js";
+import { loadKTX2 } from "../../Source/Cesium.js";
+import KTX2Transcoder from "../../Source/Core/KTX2Transcoder.js";
```
```suggestion
import { KTX2Transcoder } from "../../Source/Core/KTX2Transcoder.js";
```

lilleyse replied:
whoops sorry for the incomplete fix (CC https://github.com/CesiumGS/cesium/pull/9513/commits/938a742928af2701b410149a99c4e3b081ec1d48)

lilleyse reviewed (commented):
Feedback on tests

ebogo1 commented on Specs/Core/loadKTX2Spec.js:
```diff
@@ -0,0 +1,295 @@
+        fail();
+      })
+      .otherwise(function (error) {
+        expect(error).toBeDefined();
```
@lilleyse This throws a 404 in Sandcastle but an empty error object in spec... I thought it might be better to add `expect(error.statusCode).toBe(404)` or something similar but wasn't sure if this was possible.

lilleyse replied:
Hm yeah I'm seeing that too. It doesn't need to hold up the PR.

ebogo1 commented:
Should be good for final review!

lilleyse commented:
One of the KTX2 models isn't working in Sandcastle
> Specs/Data/Models/Box-Textured-KTX2-Basis/CesiumTexturedBoxTest.gltf

![image](https://user-images.githubusercontent.com/915398/123518450-faaa9b00-d673-11eb-8104-5996801fd487.png)


It's because the texture is 400x400 and the default sampler uses `REPEAT` which requires power-of-two textures in WebGL 1. We don't have a way of turning compressed textures into power-of-two textures like we do for regular textures. It works in WebGL 2 which doesn't have the same restriction: [Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=tVZtb9s2EP4rrD8pgEPZTtJ2nhMssbe1aIN2tfsC1MVASyeLCEVqJOUkLfLfdyQlW3byxWhtBIF4d3zunoc8kiumyYrDLWhyTiTckjEYXhX0k7dF807ix2MlLeMS9LzTJT/mkpAELXBn35WWK2mGwUiIhv8qMPYzLJZiMCRWV+AcD133n8tMXam7IcmYMOBNBgQkDuK1THnCrNJb3pyl6tYEnNqiKpFeSl4wC2v7w9HvczmXWSU9Fkk0oPtapSCiSosuyYEvc3sUqgx8KUjLLQdDNRRqBZdCRAEGI1CVUhnuwc4bTcZMW/xi8oRmWhUTWGoAEwXix/3BCe29OD193v+tG0ynp7R31jt50XteG0IV7tslCmlyYCmXy02Wa2ZzatUHNDNpov7J2Sa45DbJMbS3tmglxJYhL3dW8lVI8N5N/YDRUZ2xG9C6HmKTQmmOwrBt5jONpWRKF4bmO3D/VKi0lhheC9Ho1nAudUO4SeGFv0f03YVgaRrV+0iyApfXrd026nAXv1XvsD2o3YXbA+vdSRCRt2ExgONeqor3/A7ElH/HpP3By42X3TnvNGECPYMe/mpf2NEPtXCBiNUsuYH0z4ZfIIoRD468Vze0C/q+unl1Wa6RhmTeueQ60Syz2GQ1ORn6A3ui2drR0YZMe5s3NoI4lMbhb1pCYuIJsyz2USbG9jueYbpKQ3r8ZvZlcHzFDDdxWOjGg1Ez7GK6FDZbF+N+ZygA7TXjQH6tRfcJShOtJPwaPlNWlAI8mSKQCUX7FO1vLHuxVXX/bL+i/9aqkin5BDlPxMGqD1nqJNujRwz2qf6VsgR3ErliQiglDyt+nWR79HgB9t0211zckJmukpvDlu/y+DS7459agekNlxJSMs4ZngjW35qHJMEa/f/Fz5+qfIL1KjJWRanBGKTgazlU+T7bJtkTa7Bz/OxDxR1ubSZ7tgPPSPSsvv7+Qlp4LE7AhrcKNVVZKm2NPzuj+vQ3CUg4akEQcsslvl8oXh7athRxmsxybshC4+sGn16pAkMk9m0NTDww+Sj5CrRhAl9bax42nNFm3tkANnKgFL+knZ12jy17t/U3f+lPmUwTZiwea3jBz5QSC6avQVZRfRu6iZ1uZ2TsvYCLBu4PXngp8LKOsGILWDEyMvECtwZYmhjTZBzF7amjlK8IT8+feLuSRDBj0JNVwt/2887FKMb4R1OF8s+cd6i/YPcuLO9fvA1GSukoxuHTM20guIP8Pw)

![image](https://user-images.githubusercontent.com/915398/123518562-655bd680-d674-11eb-8eb6-2713cfc87232.png)

[KHR_texture_basisu](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_texture_basisu/README.md) doesn't have a power-of-two requirement but it does recommend it for maximum compatibility.

We should either

* Update the test models to be 256 x 256
* Change the wrap mode to `CLAMP_TO_EDGE` for non-power-of-two compressed textures.

The second option isn't great if the model is supposed to use repeat texturing, like a brick texture repeated across the side of a building. I think the first option is better for now. We should encourage power-of-two compressed textures in the Model.js documentation and explain why. We should also print a warning when this situation happens.

Another problem I noticed is that if the texture is power-of-two but the sampler requires a mipmap (like if `minFilter` is `9986` aka `NEAREST_MIPMAP_LINEAR`) and the texture does not have a mipmap it will appear black as well. This is discouraged (but not disallowed) in `KHR_texture_basisu`

> When a texture refers to a sampler with mipmap minification or when the sampler is undefined, the KTX image SHOULD contain a full mip pyramid

The solution here would be to set the `minFilter` to either `TextureMinificationFilter.NEAREST` or `TextureMinificationFilter.LINEAR` like [here](https://github.com/CesiumGS/cesium/blob/4fb9fbb097663898733ef5850f4df7d0174f3af0/Source/Scene/Model.js#L2880-L2892). There's also a mistake with that code - `textureMinificationFilter` should be `minificationFilter` and `textureMagnificationFilter` should be `magnificationFilter`.

Make the same fixes in `GltfLoaderUtil.createSampler` which is used by the new model loading code.

![image](https://user-images.githubusercontent.com/915398/123519016-cf757b00-d676-11eb-8a4c-cf7dda499d86.png)

ebogo1 commented on Source/Scene/GltfLoaderUtil.js:
```diff
@@ -104,6 +106,21 @@ GltfLoaderUtil.createSampler = function (options) {
+    }
+  }
+
   if (
```
Should the `usesTextureTransform` be an else-if here? Not sure how the extension would interact with KHR_texture_basisu.

lilleyse replied:
They are orthogonal to each other so what you have is fine. Though it could all be one if statement:

```js
  if (
    (compressedTextureNoMipmap || usesTextureTransform) &&
    minFilter !== TextureMinificationFilter.LINEAR &&
    minFilter !== TextureMinificationFilter.NEAREST
  ) {
    if (
      minFilter === TextureMinificationFilter.NEAREST_MIPMAP_NEAREST ||
      minFilter === TextureMinificationFilter.NEAREST_MIPMAP_LINEAR
    ) {
      minFilter = TextureMinificationFilter.NEAREST;
    } else {
      minFilter = TextureMinificationFilter.LINEAR;
    }
  }
```

ebogo1 commented on Source/Scene/GltfLoaderUtil.js:
```diff
@@ -74,6 +75,7 @@ GltfLoaderUtil.createSampler = function (options) {
   options = defaultValue(options, defaultValue.EMPTY_OBJECT);
   var gltf = options.gltf;
   var textureInfo = options.textureInfo;
+  var compressedTextureNoMipmap = options.compressedTextureNoMipmap;
```
My idea was that whoever calls `createSampler` would follow the logic: 

```js
if (textureIsCompressed && textureHasNoMipmap) {
  // make a note to pass `compressedTextureNoMipmap = true` to createSampler()
}
```

lilleyse replied:
You can add code like that to `GltfTextureLoader` now. It should already know whether the texture is compressed and whether it doesn't have a mipmap.

ebogo1 commented on Source/Scene/Model.js:
```diff
@@ -1366,6 +1371,8 @@ function containsGltfMagic(uint8Array) {
  * </li><li>
  * {@link https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_texture_transform/README.md|KHR_texture_transform}
+ * </li><li>
+ * {@link https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_texture_basisu/README.md|KHR_texture_basisu}
```
I don't think there was a reason for this to only be in the constructor docs and not `fromGltf` docs?

ebogo1 reviewed (commented):
Good finds - I responded to both in 4325679. Had a couple questions -

lilleyse commented on Source/Scene/GltfLoaderUtil.js:
```diff
@@ -66,6 +66,7 @@ GltfLoaderUtil.getImageIdFromTexture = function (options) {
  * @param {Object} options Object with the following properties:
  * @param {Object} options.gltf The glTF JSON.
  * @param {Object} options.textureInfo The texture info object.
+ * @param {Object} [options.compressedTextureNoMipmap] Whether the texture is compressed and has an embedded mipmap.
```
Some suggestions: changed `{Object}` to `{Boolean}`, added a default value (the default value should also be set in the code below with `defaultValue`), and updated the description.

```
 * @param {Boolean} [options.compressedTextureNoMipmap=false] True when the texture is compressed and does not have an embedded mipmap.
```

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -180,6 +180,11 @@ var uriToGuid = {};
  * </ul>
  * </p>
  * <p>
+ * Note: for models with compressed textures using the KHR_texture_basisu, we recommend power of 2 textures in both dimensions
```
"using the KHR_texture_basisu" -> "using the KHR_texture_basisu extension"

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -2893,11 +2900,46 @@ function createTexture(gltfTexture, model, context) {
+    // Warning to encourage power of 2 texture dimensions with KHR_texture_basisu
+    if (npot && PixelFormat.isCompressedFormat(internalFormat)) {
+      console.warn(
+        "For maximim compatibility, we encourage power of 2 dimensions when using compressed textures in glTF. See the Model.js constructor documentation for more information."
```
A similar warning should go in `GltfTextureLoader`

lilleyse replied:
The warning should only be printed if the wrap mode is `REPEAT` or `MIRRORED_REPEAT`, since `CLAMP_TO_EDGE` is totally ok.

Also only print the error for WebGL 1. You can check if a WebGL 2 context is used with `context.webgl2`

lilleyse replied:
The warning should be a bit more serious and informative. Something like:

"Compressed texture uses REPEAT or MIRRORED_REPEAT texture wrap mode and dimensions are not powers of two. The texture may be rendered incorrectly."

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -2893,11 +2900,46 @@ function createTexture(gltfTexture, model, context) {
+      );
+    }
+
+    if (
+      !defined(gltfTexture.mipLevels) &&
+      (minFilter === TextureMinificationFilter.NEAREST_MIPMAP_NEAREST ||
+        minFilter === TextureMinificationFilter.NEAREST_MIPMAP_LINEAR)
+    ) {
+      sampler = new Sampler({
+        wrapS: sampler.wrapS,
+        wrapT: sampler.wrapT,
+        minificationFilter: TextureMinificationFilter.NEAREST,
+        magnificationFilter: sampler.magnificationFilter,
+      });
+    } else if (
+      !defined(gltfTexture.mipLevels) &&
+      (minFilter === TextureMinificationFilter.LINEAR_MIPMAP_NEAREST ||
+        minFilter === TextureMinificationFilter.LINEAR_MIPMAP_LINEAR)
+    ) {
+      sampler = new Sampler({
+        wrapS: sampler.wrapS,
+        wrapT: sampler.wrapT,
+        minificationFilter: TextureMinificationFilter.LINEAR,
+        magnificationFilter: sampler.magnificationFilter,
+      });
+    }
```
I feel like this code could be more concise. Maybe `minificationFilter` can get set first, and then the sampler can get created.

lilleyse commented on Specs/Scene/GltfLoaderUtilSpec.js:
```diff
@@ -145,7 +145,7 @@ describe("Scene/GltfLoaderUtil", function () {
   });
 
-  it("createSampler throws if gltf is undefined", function () {
+  it("createSampler throws if textureInfo is undefined", function () {
```
Good catch :+1:

ebogo1 commented:
@lilleyse Think I got everything with the `GltfLoader` changes.

lilleyse commented:
Looks good. I just modified the `GltfTextureLoaderSpec` test to piggy back off the existing glTF. I also remembered to update the models in the `GltfLoader` folder so that they are also 256x256.

Should be ready to merge after we update the hosted environment map.

ebogo1 commented:
Thanks. The ibl file has been updated.

lilleyse commented:
Thanks @ebogo1. I pushed what I hope is the last commit.

I realized I forgot to change `400` to `256` in the `GltfLoader` tests. Also one of the texture tests was giving a slightly different result so I changed it to use an epsilon comparison. See https://github.com/CesiumGS/cesium/pull/9513/commits/5b45436ea680c054dabd1e7e4a9c9a3280ddfadc.

Will merge after CI finishes.

lilleyse commented:
Merged. Thanks for the awesome work here @ebogo1!

lilleyse commented:
@ebogo1 I noticed an issue with IE11 that we should fix before the July release.

`KTX2Transcoder.js` is automatically pulling in `transcodeKTX2` which pulls in `parseKTX2` which pulls in `ktx-parse.modern.js` which uses non ES5 features (which we were aware of https://github.com/CesiumGS/cesium/pull/9513#discussion_r652287683)

This should be fixable by wrapping the `_transcodeTaskProcessor` creation in a getter like what `DracoLoader` does. That way apps that don't use KTX2 will still function in IE11 for another month.

ebogo1 commented:
For anyone else who reads this in the future, I opened https://github.com/CesiumGS/cesium/pull/9650 to fix the IE error.
